src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -2702,14 +2700,22 @@ private void PushOne()
                         }
 
                         Push(_stage.Out!, result.Value);
-                        if (Todo < _stage._parallelism && !HasBeenPulled(inlet))
-                            TryPull(inlet);
+                        PullIfNeeded();
                     }
 
                     break;
                 }
             }
 
+            private void PullIfNeeded()
+            {
+                if(IsClosed(_stage.In) && _buffer!.IsEmpty)
+                    CompleteStage();
+                else if(_buffer!.Used < _stage._parallelism && !HasBeenPulled(_stage.In))
+                    TryPull(_stage.In);
+                // else already pulled and waiting for next element
+            }
+
```

Comment:
All `Pull` and `TryPull` is replaced with this code, this guarantees that we complete the stage automatically if upstream is completed and we drained our buffer.